### PR TITLE
Add DimensionSelector id -> X caches.

### PR DESCRIPTION
**Pull request by gianm**
Adds the method DimensionSelectorUtils.cacheIfPossible to aid with
creation of caches for evaluating functions of dimension dictionary IDs.
Uses an array cache when possible, an LRU cache when it seems like it
might be a good idea, and skips caching otherwise.

The caches are used in three places in this patch:

- Cardinality aggregator (caches hashes)
- Single-dimension-input expression selector (caches expression results)
- Generic expression selector (caches name for any dimension inputs)

Benchmarking showed that both caches help when there are repeated values,
but also the LRU cache could introduce unwanted overhead for dimensions
with few repeating values, so this patch implements two mitigations:

- Skip caching completely if cardinality is >90% of the row count.
- After iterating through a few multiples of the cache size, examine
  the cache hit rate and decide to either freeze it (the cache becomes
  read-only) or ignore it (the cache is cleared and future operations
  are uncached).

The caches aim to be 1MB or less per column.

### CODE REVIEW

File: processing/src/main/java/io/druid/query/dimension/ArrayCacheIntFunction.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+package io.druid.query.dimension;
+
+import java.util.function.IntFunction;
+
+/**
+ * Array cache for an IntFunction, intended for use with DimensionSelectors.
+ *
+ * @see io.druid.segment.DimensionSelectorUtils#cacheIfPossible
+ */
+public class ArrayCacheIntFunction<T> implements IntFunction<T>
```
[the remark this answers is not in the document]
Comment (leventov) [reply processing/src/main/java/io/druid/query/dimension/ArrayCacheIntFunction.java]: Maybe just call it "ArrayDimensionCache" (and "LruDimensionCache") with a method like "getOrCompute()", and then return method reference in `DimensionSelectorUtils.cacheIfPossible()`

File: processing/src/main/java/io/druid/query/dimension/LruCacheIntFunction.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * LRU cache for an IntFunction, intended for use with DimensionSelectors.
+ *
+ * @see io.druid.segment.DimensionSelectorUtils#cacheIfPossible
+ */
+public class LruCacheIntFunction<T> implements IntFunction<T>
+{
+  // After a certain point, freeze or ignore the cache, based on hit rate. The idea is that hopefully by then,
+  // LRU should have some reasonable values in the cache if there are reasonable values to be found.
+  private static final int CACHE_FREEZE_FACTOR = 10;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/dimension/LruCacheIntFunction.java]: How many passes over a column are done in total?

File: processing/src/main/java/io/druid/segment/virtual/SingleStringInputCachingExpressionColumnValueSelector.java [hunk 3]
```diff
@@ -117,48 +117,15 @@ private ExprEval eval()
...
-      if (value == null) {
-        value = expression.eval(bindings);
-        m.putAndMoveToFirst(id, value);
-
-        if (m.size() > CACHE_SIZE) {
-          m.removeLast();
-        }
+      return singleValueEvalCache.apply(row.get(0));
+    } else {
+      // Tread non-singly-valued rows as nulls, just like ExpressionSelectors.supplierFromDimensionSelector.
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/virtual/SingleStringInputCachingExpressionColumnValueSelector.java]: Treat

**Comment by leventov** [comment]
Also please fix "unused declaration", IntelliJ inspection fails

**Comment by gianm** [comment]
Closing this, I don't have bandwidth to continue working on it at this time, and it conflicts with #6794 which is more important to me anyway. Will have to consider reviving in the future.

**Pull request closed without merging**